src/openforms/submissions/api/viewsets.py

```diff
@@ -233,6 +234,13 @@ def _complete(self, request, *args, **kwargs):
         ---
         """
         submission = self.get_object()
+
+        update_serializer = SubmissionUpdateSerializer(
+            instance=submission, data=request.data
+        )
+        update_serializer.is_valid(raise_exception=True)
+        update_serializer.save()
```

Comment:
I don't really understand why we need a separate serializer for this? IMO we can bake this entirely into the existing flow, but we'll need to refactor that a little bit.

The basis should be the `CompletionValidationSerializer` for this entire request-response cycle, in essence:

```py
serializer = SubmissionCompletionSerializer(
    instance=submission,
    data=request.data,
)
serializer.is_valid(raise_exception=True)
...
serializer.save()
return Response(serializer.data)
```

Where `SubmissionCompletionSerializer` is refactored to handle the behaviour of validation + succesful serialization:

1. Validate that privacy policy is accepted
2. Validate that required steps are done etc.

You an do that by marking serializer fields as read/write only so that they are only emitted in response or request body.

This does not need to be a model serializer, there is nothing stopping you from implementing the `save` method yourself and then just adding a line `self.instance.privacy_policy_accepted = True` and `self.instance.save()` to it.

Reply:
@sergei-maertens I tried implementing this suggestion, I am not confident that it now matches what you had in mind :sweat_smile: 
The coverage is failing I think because `NestedSubmissionRelatedField` is no longer used anywhere. 
If these changes are going in the right direction then I'll remove it